### PR TITLE
First pass at transaction support

### PR DESCRIPTION
**Pull request by jskeet**
No integration tests yet, as I'm trying to find out how to reliably force a retriable failure.
The rules about when to roll back and which commits can be retried are slightly up in the air, but
we can change them later. Right now:

- We roll back if we haven't even tried to commit
- We roll back if we tried to commit and it failed with abort/cancellation/deadline
- We retry the whole transaction if commit failed with abort

### CODE REVIEW

File: apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs [hunk 1]
```diff
@@ -215,5 +217,129 @@ string ExtractPath(DocumentReference documentReference)
...
+        /// </summary>
+        /// <typeparam name="T">The result type of the callback.</typeparam>
+        /// <param name="callback">The callback to execute. Must not be null.</param>
+        /// <param name="options">The options for the transaction. May be null, in which case default options will be used.</param>
+        /// <param name="cancellationToken">A cancellation token for the operation. This is exposed to the callback through <see cref="Transaction.CancellationToken"/>
+        /// and applied to all RPCs to begin, rollback or commit the transaction.</param>
+        /// <returns>A task which completes when the transaction has committed. The result of the task then contains the result of the callback.</returns>
+        public async Task<T> RunTransactionAsync<T>(Func<Transaction, Task<T>> callback, TransactionOptions options = null, CancellationToken cancellationToken = default)
+        {
+            ByteString previousTransactionId = null;
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs]: Check callback isn't null

File: apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs [hunk 2]
```diff
@@ -215,5 +217,129 @@ string ExtractPath(DocumentReference documentReference)
...
+        /// <typeparam name="T">The result type of the callback.</typeparam>
+        /// <param name="callback">The callback to execute. Must not be null.</param>
+        /// <param name="options">The options for the transaction. May be null, in which case default options will be used.</param>
+        /// <param name="cancellationToken">A cancellation token for the operation. This is exposed to the callback through <see cref="Transaction.CancellationToken"/>
+        /// and applied to all RPCs to begin, rollback or commit the transaction.</param>
+        /// <returns>A task which completes when the transaction has committed. The result of the task then contains the result of the callback.</returns>
+        public Task<T> RunTransactionAsync<T>(Func<Transaction, T> callback, TransactionOptions options = null, CancellationToken cancellationToken = default)
+        {
+            GaxPreconditions.CheckNotNull(callback, nameof(callback));
+            return RunTransactionAsync(async transaction => callback(transaction), options, cancellationToken);
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs]: Is there any advantage of `async transaction => ...` to ` transaction => Task.FromResult(...)` ?
The later seems to me to be easier to understand! And won't give CS1998 warning.

File: apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs [hunk 3]
```diff
@@ -215,5 +217,129 @@ string ExtractPath(DocumentReference documentReference)
         internal IAsyncEnumerable<CollectionReference> ListCollectionsAsync(DocumentReference parent) =>
             Client.ListCollectionIdsAsync(parent?.Path ?? RootPath)
                 .Select(id => new CollectionReference(this, parent, id));
+
+        // TODO: Is it appropriate to use ConfigureAwait(false) here? Should the caller be able to specify the context for executing their callback?
+        // TODO: Collect up all the exceptions we've retried?
+        // TODO: Is it nice to have all these overloads?
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs]: I think just keeping the two overloads with async callbacks is fine. As the `RunTransactionAsync` method is itself async, I think requiring an async callback is reasonable.

File: apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data.Tests/FirestoreDbTest.RunTransactionAsync.cs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+                    await transaction.GetDocumentSnapshotAsync(db.Document("col/x"));
+                    throw new IOException("Bang!");
+                }));
+            Assert.Equal("Bang!", exception.Message);
+
+            Assert.Empty(client.CommitRequests);
+            Assert.Equal(new[] { CreateRollbackRequest("transaction 1") }, client.RollbackRequests);
+        }
+
+        /*[Theory]
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data.Tests/FirestoreDbTest.RunTransactionAsync.cs]: Remove?

Comment (jskeet) [reply apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data.Tests/FirestoreDbTest.RunTransactionAsync.cs]: Whoops - reinstated instead. Changed it to a fact to make debugging work.

File: apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs (hunk 3 above)
Comment (jskeet) [reply apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs]: I thought so until I started writing tests - but as the modification methods are synchronous, I think it won't be uncommon to have a synchronous callback, in which case you get a warning due to an async lambda with no awaits. (Or you have to return Task.FromResult directly.)

File: apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs]: Done.

File: apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs (hunk 2 above)
Comment (jskeet) [reply apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/FirestoreDb.cs]: It was for the sake of propagating exceptions in the same way - but as we're awaiting the result immediately, it should be fine actually. Will change it.

**Pull request merged**